### ISSUE TITLE
Groovy: Reflection causes exceptions after several runs

### ISSUE THREAD

**Issue by spinscale** [elastic/elasticsearch#16536]
A user in the [watcher forum](https://discuss.elastic.co/t/failed-to-execute-watch-transform-noclassdeffounderror/41257/5) hit a `ClassNotFoundError`, after a watch ran successful several times.

Turns out this is a groovy issue and can be reproduced using this test:

```java
package org.elasticsearch.script.groovy;

import org.elasticsearch.common.settings.Settings;
import org.elasticsearch.script.CompiledScript;
import org.elasticsearch.script.ExecutableScript;
import org.elasticsearch.script.ScriptService;
import org.elasticsearch.test.ESTestCase;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class GroovyCompilationTests extends ESTestCase {

    private GroovyScriptEngineService se;

    @Override
    public void setUp() throws Exception {
        super.setUp();
        se = new GroovyScriptEngineService(Settings.EMPTY);
        // otherwise will exit your VM and other bad stuff
        assumeTrue("test requires security manager to be enabled", System.getSecurityManager() != null);
    }

    @Override
    public void tearDown() throws Exception {
        se.close();
        super.tearDown();
    }

    public void testThatPotentialBugWithSeveralExecutions() throws Exception {
        String script = "return hits.collect({\"${it.message}\"})";

        Map<String, Object> data = new HashMap<>();
        data.put("message", "The big lebowski");

        List<Object> hitsArray = new ArrayList<>();
        hitsArray.add(data);

        Map<String, Object> hits = new HashMap<>();
        hits.put("hits", hitsArray);

        Object compiledObject = se.compile(script, Collections.emptyMap());
        CompiledScript compiledScript = new CompiledScript(ScriptService.ScriptType.INLINE, "test", "groovy", compiledObject);
        ExecutableScript executable = se.executable(compiledScript, hits);

        // 100 runs will fail, 10 runs work..
        for (int i = 0; i < 100; i++) {
            executable.run();
        }
    }

}
```

@ywelsch already found out, that one can trigger this immediately by setting the `-Dsun.reflect.noInflation=true` property.

For more info about that, check out https://blogs.oracle.com/buck/entry/inflation_system_properties

**Comment by ywelsch** [comment]
The reason it fails is that after 15 iterations inflation kicks in. An explanation of inflation is given here:
https://blogs.oracle.com/buck/entry/inflation_system_properties
The following stackoverflow post also has some information:
http://stackoverflow.com/questions/20527776/why-does-my-custom-securitymanager-cause-exceptions-the-16th-time-i-create-an-ob

Adding the following permission to plugin-security.policy solves it:
```
permission org.elasticsearch.script.ClassPermission "sun.reflect.MethodAccessorImpl";
```

**Comment by ywelsch** [comment]
@rmuir would it make sense to add `sun.reflect.ConstructorAccessorImpl` and `sun.reflect.MethodAccessorImpl` to `STANDARD_CLASSES` in `ClassPermission`?
I'm not sure whether other script plugins might be affected by this issue.

**Comment by rmuir** [comment]
The issue may be groovy specific. Groovy is the only one that tries to access sun.reflect directly, hence the only one with this:
```
  permission java.lang.RuntimePermission "accessClassInPackage.sun.reflect";
```

**Comment by rmuir** [comment]
Also note that groovy runtime already has ConstructorAccessor, because only a constructor was reflected in a loop:
```
  permission org.elasticsearch.script.ClassPermission "sun.reflect.ConstructorAccessorImpl";
```
So i would add the other inflation-related class right there beside it, just for groovy. We should add a loop test for the case. If we can add the same loop test to Javascript/Python to show it works fine without this, that would be nice.

**Comment by gkop** [comment]
@spinscale trying to understand the quick fix in the meantime before 2.1.1 is out.

You said at https://discuss.elastic.co/t/failed-to-execute-watch-transform-noclassdeffounderror/41257/8 :

> in the meantime, you can rewrite your code, so it does not use closures, so everything should be fine

Is the problem really with using closures? My experimentation suggests the issue is the use of Groovy's default `it` variable, and the problem goes away when I name the variable. If this is correct it's a very easy fix indeed.

Maybe we're saying the same thing - I'm new to Groovy - but would you please clarify?
